### DISCUSSION TITLE
try/catching mrgsim errors

### DISCUSSION THREAD

**Discussion by efernandezbi** [metrumresearchgroup/mrgsolve#1120] [Q&A]
Hi,
Using mrgsim with idata is a great way to produce virtual patient populations. I use expand.idata to create parameter sets randomly distributed and produce VP, but there is always a risk that some sets lead to mrgsim/lsoda failing to converge. Example:

```
### pop ###
np = 1000
idataCL <- expand.idata(CL=rnorm(np, 4.1, 75*4.1/100))
idataV1 <- expand.idata(Central=rnorm(np, 5.25, 35*5.25/100))
idataQ <- expand.idata(Q=rnorm(np, 11, 30*11/100))
idataV2 <- expand.idata(Peripheral=rnorm(np, 470, 175*470/100))
idataTc <- expand.idata(Tc=rnorm(np, 100000, 25*100000/100))
idataCc <- expand.idata(Cc=rnorm(np, 100000, 30*100000/100))

data <- idataCL %>%
  left_join(idataV1, by="ID") %>%
  left_join(idataQ, by="ID") %>%
  left_join(idataV2, by="ID") %>%
  left_join(idataTc, by="ID") %>%
  left_join(idataCc, by="ID")

endtime = 1344 
sim <- mod %>%
  ev(amt=50e-6, ii=168, addl=7) %>%
  idata_set(data) %>%
  mrgsim(end=endtime, atol = 1E-20)
```

With 1000 patients, I sometimes get:

```
[mrgsolve] lsoda returned with negative istate: -4
  repeated error test failures (check all inputs).

Error: simulation terminated.
```

I have some workarounds like decreasing atol, however this sometimes work but this is not 100% failproof, and need to run the script a few times until it succeeds. Is there any possibility to catch mrgsim errors and ask it to drop those simulations that fail and skip to next patient? Thanks a lot for any advice.

**Comment by kylebaron** [comment]
Hi @efernandezbi  -  

I don't believe there is ability to try and catch a problem with an individual subject from _within the model_. 

You could split the data and run one at a time and wrap that in `try()` 

```r
sim <- try(
   mod %>%
   ev(amt=50e-6, ii=168, addl=7) %>%
   idata_set(first_subject) %>%
   mrgsim(end=endtime, atol = 1E-20)
)
```

This will probably be slower; maybe batching this up by 10s would help? Not sure. 

I'm not sure what is the model structure, but I might look a little bit at why some parameters are failing. It looks like the dose is very small and you are having to increase `atol` quite a bit. Maybe there is something that can be addressed in how the model is set up. But again I'd have to look at it first to say that for sure. 

Best regards,
Kyle

**Comment by efernandezbi** [comment]
Thank you, I will try this. Dose is small but parameters are scaled accordingly and the deterministic model runs fine. Indeed I could loop over the data set and run simulations one by one. An alternative solution to try is possibly() but I am unsure I can use it to wrap mrgsim().

**Comment by FelicienLL** [comment]
Hi @efernandezbi, 
With your approach, you are currently generating negative parameter numbers. I would advise you to define inter-individual variability as proportional or exponential such as: 
``` r
CL <- rnorm(1000, 4.1, 75*4.1/100)
any(CL < 0)
#> [1] TRUE
CL <- 4.1 * exp(rnorm(1000, sd = 75/100))
any(CL < 0)
#> [1] FALSE
```
In addition, if you have very large variability, assuming independent distributions might lead to the generation of unlikely combinations of, e.g., clearance and volume, hence increasing the risk of generating "hard to simulate" profiles and the solver to fail. Sampling from a multivariate distribution could help sometimes, see `MASS::mvrnorm()` or `mrgsolve::mvgauss()`

Best regards
Félicien

**Comment by efernandezbi** [comment]
> Hi @efernandezbi, With your approach, you are currently generating negative parameter numbers.

Dear Félicien,

Many thanks for your answer and advice. Indeed there is a large spread and possibility of generating negative values, but I was trying to reproduce the simulations from a published article, with same methods and variance. But you are right this would need to be addressed.
